.NET Guide - ReSharper Essentials (#153)
JetBrains/jetbrains_guide: sites/dotnet-guide/contents/tutorials/resharper-essentials/shortcuts-editing-completion/index.md

---
type: TutorialStep
date: 2021-03-04
title: Shortcuts, Editing and Code Completion
technologies: [.net, csharp]
topics: [completion, editing, inspections, navigation]
products: [resharper]
author: matkoch
subtitle: Let's dive into the most important productivity shortcuts, editing, and code completion.
thumbnail: ./thumbnail.png
longVideo: 
    poster: ./poster.png
    url: https://TODO
---

### Get Going with ReSharper and <kbd>Alt-Enter</kbd>

> As with any other new tool, it's very helpful to learn about its basic functionality first. ReSharper makes this very easy by providing <kbd>Alt-Enter</kbd> as the single entry point to [find actions and options](https://www.jetbrains.com/help/resharper/Navigating_to_Action.html), or to [apply fixes](https://www.jetbrains.com/help/resharper/Code_Analysis__Quick-Fixes.html) to our code.

Reviewer: `easy` -> `straightforward`

> After pressing <kbd>Alt-Enter</kbd>, we can start typing to find our desired action, which can include navigation methods, refactorings, or one of the several tool windows that ReSharper has. For better learning, the list of actions always shows the assigned keyboard shortcut as well.

Reviewer: ```suggestion
After pressing <kbd>Alt-Enter</kbd>, we can start typing to find our desired action, which can include navigation methods, refactorings, or one of the several ReSharper tool windows. The list of actions always shows the assigned keyboard shortcut to help us learn faster.
```

Having <kbd>Alt-Enter</kbd> makes it very easy to keep your hands on the keyboard at all time, so you can avoid having to switch between input devices all the time.

### Editing Code and Code Completion

> ReSharper can help you while writing code. Whatever your caret position may be, ReSharper has good selection of [code completion](https://www.jetbrains.com/help/resharper/Auto-Completing_Code.html) items for us!

Reviewer: ```suggestion
ReSharper can help you while writing code. Where ever your caret position may be, ReSharper has a good selection of [code completion](https://www.jetbrains.com/help/resharper/Auto-Completing_Code.html) items for us!
```

For instance, in a class body, we would get all possible access modifiers, whereas in function bodies we get all control flow keywords, imported types, parameters and variables available in that particular scope. Through [double completion](https://www.jetbrains.com/help/resharper/Coding_Assistance__Code_Completion__Double.html) we can even filter the list to show only relevant items, for instance to show only items of type `bool`.

Code completion also works for other languages and file formats, and even inside strings! Using [language injections](https://www.jetbrains.com/help/resharper/Auto-Completing_Code.html), HTML, regular expressions, and JSON is highlighted in strings, and shows relevant tags, character classes, or property names when needed.

> ### Generating Code

Reviewer 2: Replace that part with:

### Code Templates

ReSharper's code templates can generate otherwise tedious code for you! The most basic templates are [live/surround templates](https://www.jetbrains.com/help/resharper/Templates__Template_Basics__Template_Types.html) and [file templates](https://www.jetbrains.com/help/resharper/Creating_a_File_Template.html) (even with [multiple files](https://www.jetbrains.com/help/resharper/Templates__Creating_and_Editing_Templates__Multifile.html)). A lot of general-purpose templates come preinstalled, for instance the `cw` template for `Console.WriteLine`, the `nguid` template to generate a new `Guid`, or the `try` surround template to generate a `try-catch` block (even works by temporarily overwriting the selected code). Templates can have placeholders, which can be entered when applying a template. But these placeholders can also make use of the [wide range of macros](https://www.jetbrains.com/help/resharper/Template_Macros.html) to give suggestions, including the current user, date and time, expected types for generic arguments, or transformation of other macro values (identifier name from text).

Another kind are [postfix templates](https://www.jetbrains.com/help/resharper/Postfix_Templates.html). These are typically written in a reverse order of thinking. For instance, instead of using the `foreach` live template and then defining the `numbers` collection as we go through the placeholders, with a postfix template we can write `numbers.foreach` and ReSharper will complete what we want with much less effort. A [full list of postfix templates](https://www.jetbrains.com/help/resharper/Postfix_Templates.html#list) is available from the options dialog. Postfix templates can't be defined in a simple way like live or surround templates (they require to use the ReSharper SDK), however, ReSharper also allows to define [source templates](https://www.jetbrains.com/help/resharper/Source_Templates.html) alongside our code. These can be used in a similar manner with just some smartness limitations.

Reviewer 2: That too much?

Author: Commit incoming!

> ReSharper's code generation actions can create a lot of code for you! Constructors to initialize any set of type members, methods and properties required by an interface you're implementing, equality checks, formatting members, and more.

Reviewer: ```suggestion
ReSharper's code generation actions can generate otherwise tedious code for you! Constructors to initialize any set of type members, methods, and properties required by an interface you're implementing, equality checks, formatting members, and more.
```

I felt like "a lot of code" didn't sound like it was helpful but might be considered _junk_.
